File: Byte_Compressor/src/core/bitstream.js
/*
 * ByteCompressor - Bitstream I/O (Node.js implementation)
 * Bit-level read/write operations for entropy coding
 */

class BitStream {
    constructor(bufferOrSize) {
        if (typeof bufferOrSize === 'number') {
            this.data = Buffer.alloc(bufferOrSize);
            this.capacity = bufferOrSize;
        } else {
            this.data = Buffer.from(bufferOrSize);
            this.capacity = this.data.length;
        }
        this.bytePos = 0;
        this.bitPos = 0;
        this.totalBits = 0;
    }

    static forWrite(capacity) {
        const bs = new BitStream(capacity);
        return bs;
    }

    static forRead(buffer, totalBits) {
        const bs = new BitStream(buffer);
        bs.totalBits = totalBits || buffer.length * 8;
        return bs;
    }

    writeBits(value, count) {
        for (let i = count - 1; i >= 0; i--) {
            if (this.bytePos >= this.capacity) return -1;
            if ((value >> i) & 1) {
                this.data[this.bytePos] |= (1 << (7 - this.bitPos));
            }
            this.bitPos++;
            this.totalBits++;
            if (this.bitPos === 8) {
                this.bitPos = 0;
                this.bytePos++;
            }
        }
        return 0;
    }

    readBits(count) {
        let value = 0;
        for (let i = count - 1; i >= 0; i--) {
            const absBit = this.bytePos * 8 + this.bitPos;
            if (absBit >= this.totalBits) return -1;
            if (this.data[this.bytePos] & (1 << (7 - this.bitPos))) {
                value |= (1 << i);
            }
            this.bitPos++;
            if (this.bitPos === 8) {
                this.bitPos = 0;
                this.bytePos++;
            }
        }
        return value >>> 0;
    }

    bitsWritten() { return this.totalBits; }
    bytesUsed() { return Math.ceil(this.totalBits / 8); }

    flush() {
        if (this.bitPos > 0) {
            this.bytePos++;
            this.bitPos = 0;
        }
    }
}

module.exports = { BitStream };

File: Byte_Compressor/src/utils/bitstream.c
/*
 * ByteCompressor - Bitstream I/O implementation
 */

#include "bitstream.h"
#include <string.h>

void bs_init_write(BitStream *bs, uint8_t *buffer, size_t capacity) {
    bs->data = buffer;
    bs->capacity = capacity;
    bs->byte_pos = 0;
    bs->bit_pos = 0;
    bs->total_bits = 0;
    memset(buffer, 0, capacity);
}

void bs_init_read(BitStream *bs, const uint8_t *buffer, size_t total_bits) {
    bs->data = (uint8_t *)buffer;
    bs->capacity = (total_bits + 7) / 8;
    bs->byte_pos = 0;
    bs->bit_pos = 0;
    bs->total_bits = total_bits;
}

int bs_write_bits(BitStream *bs, uint32_t value, int count) {
    if (count <= 0 || count > 32) return -1;

    for (int i = count - 1; i >= 0; i--) {
        if (bs->byte_pos >= bs->capacity) return -1;

        if ((value >> i) & 1) {
            bs->data[bs->byte_pos] |= (1 << (7 - bs->bit_pos));
        }

        bs->bit_pos++;
        bs->total_bits++;
        if (bs->bit_pos == 8) {
            bs->bit_pos = 0;
            bs->byte_pos++;
        }
    }
    return 0;
}

int bs_read_bits(BitStream *bs, uint32_t *value, int count) {
    if (count <= 0 || count > 32) return -1;

    *value = 0;
    for (int i = count - 1; i >= 0; i--) {
        size_t abs_bit = bs->byte_pos * 8 + bs->bit_pos;
        if (abs_bit >= bs->total_bits) return -1;

        if (bs->data[bs->byte_pos] & (1 << (7 - bs->bit_pos))) {
            *value |= (1u << i);
        }

        bs->bit_pos++;
        if (bs->bit_pos == 8) {
            bs->bit_pos = 0;
            bs->byte_pos++;
        }
    }
    return 0;
}

int bs_write_byte(BitStream *bs, uint8_t byte) {
    return bs_write_bits(bs, byte, 8);
}

int bs_read_byte(BitStream *bs, uint8_t *byte) {
    uint32_t val;
    int ret = bs_read_bits(bs, &val, 8);
    if (ret == 0) *byte = (uint8_t)val;
    return ret;
}

size_t bs_bits_written(const BitStream *bs) {
    return bs->total_bits;
}

size_t bs_bytes_used(const BitStream *bs) {
    return (bs->total_bits + 7) / 8;
}

void bs_flush(BitStream *bs) {
    if (bs->bit_pos > 0) {
        bs->byte_pos++;
        bs->bit_pos = 0;
    }
}
